mkosi/machine.py

```diff
@@ -152,7 +157,9 @@ def __exit__(self, *args: Any, **kwargs: Any) -> None:
             self._serial.kill(signal.SIGTERM)
             self.exit_code = self._serial.wait()
             self._serial = None
-        self.stack.__exit__(*args, **kwargs)
+
+        if self.stack:
+            self.stack.__exit__(*args, **kwargs)
```

Comment:
Using "shell" we'll call `_kill()` multiple times. Since `self.stack` is only initialised once, I thought it was better to preemptively check if it exists.

Reply:
`self.stack` is not optional so I'm not sure what this if is checking. Calling `__exit__()` on the exit stack will empty the exit stack so calling it more than once isn't an issue.